src/execution_plan/ops/op_create.c

```diff
@@ -21,48 +20,43 @@ void _SetModifiedEntities(OpCreate *op) {
 
     /* For every entity within the CREATE clause see if it's also mentioned 
      * within the MATCH clause. */
+    TrieMap *matchEntities = NewTrieMap();
+    MatchClause_ReferredEntities(op->ast->matchNode, matchEntities);
+
     for(int i = 0; i < create_entity_count; i++) {
         AST_GraphEntity *create_ge;
         Vector_Get(op->ast->createNode->graphEntities, i, &create_ge);
-        
-        /* See if current entity is in MATCH clause. */
-        AST_GraphEntity *match_ge = MatchClause_GetEntity(op->ast->matchNode, create_ge->alias);
-        if(!match_ge) {
-            /* Entity is not in MATCH clause. */
-            AST_GraphEntity *ge = create_ge;
-            if(ge->t == N_ENTITY) {
-                // Node.
-                Node *n = QueryGraph_GetNodeByAlias(op->qg, ge->alias);
-                if(n->id == 0) continue;   // Node already marked for creation.
-                Node **ppn = QueryGraph_GetNodeRef(op->qg, n);
-                op->nodes_to_create[node_idx].original_node = n;
-                op->nodes_to_create[node_idx].original_node_ref = ppn;
-                n->id = 0;  /* Mark node. */
-                node_idx++;
-            } else {
-                // Edge.
-                Edge *e = QueryGraph_GetEdgeByAlias(op->qg, ge->alias);
-                if(e->id == 0) continue;   // Edge already marked for creation.
-                op->edges_to_create[edge_idx].original_edge = e;
-                op->edges_to_create[edge_idx].original_edge_ref = QueryGraph_GetEdgeRef(op->qg, e);
-                assert(QueryGraph_ContainsNode(op->qg, e->src));
-                assert(QueryGraph_ContainsNode(op->qg, e->dest));
-                op->edges_to_create[edge_idx].src_node_alias = e->src->alias;
-                op->edges_to_create[edge_idx].dest_node_alias = e->dest->alias;
-                e->id = 0;  /* Mark edge. */
-                edge_idx++;
-            }
-        }
-    }
-    
-    /* Restore IDs. */
-    for(int i = 0; i < node_idx; i++) {
-        op->nodes_to_create[i].original_node->id = INVALID_ENTITY_ID;
-    }
-    for(int i = 0; i < edge_idx; i++) {
-        op->edges_to_create[i].original_edge->id = INVALID_ENTITY_ID;
+        char *alias = create_ge->alias;
+
+        /* See if current entity needs to be created:
+         * 1. current entity is NOT in MATCH clause.
+         * 2. We've yet to accounted for this entity. */
+        if(TrieMap_Find(matchEntities, alias, strlen(alias)) != TRIEMAP_NOTFOUND) continue;
+
+        // Remember this entity to avoid duplications.
+        TrieMap_Add(matchEntities, alias, strlen(alias), NULL, TrieMap_DONT_CARE_REPLACE);
```

Comment:
TrieMap_Add returns 1 if the key is new and 0 otherwise, so you can just use that instead of also having line 34.

Reply:
Nice observations